### PR TITLE
Each commit is a bug fix

### PR DESCRIPTION
**Pull request by kc1212**

### CODE REVIEW

File: lib/master/master.go [hunk 1]
```diff
@@ -146,8 +146,8 @@ func (m *Master) Run() {
 		case beat := <-m.heartbeatChan:
 			// update task statuses
 			// check whether a job has completed all its tasks
-			m.instances[beat.addr].heartbeatChan <- true
 			m.updateOnHeartbeat(beat)
+			m.instances[beat.addr].heartbeatChan <- true
```
Comment (aaronang) [inline lib/master/master.go:150]: Could you explain why this solves the deadlock?

**Review by aaronang — APPROVED** [review]
LGTM

File: lib/master/master.go (hunk 1 above)
Comment (kc1212) [reply lib/master/master.go:150]: It's not a deadlock I don't think, the code is just stuck there for some reason. Well, the reason is that we create a new slave in `updateOnHeartbeat`, but that function is originally called after we write to `m.instances[beat.addr].heartbeatChan`. Effectively we're writing to a channel that does not exist yet. Not sure why the code is stuck there, but it works now.

**Comment by kc1212** [comment]
This is all for now, every commit is a bug fix. At the time of writing, we should be able to read status page, send jobs, and process heartbeats without locks/crashes. The next step is to fix the scheduler, sometimes the job does not get scheduled, I haven't looked into it yet.

File: lib/master/master.go (hunk 1 above)
Comment (kc1212) [reply lib/master/master.go:150]: Hence deadlock was in quotes

**Comment by victor-li** [comment]
LGTM

**Pull request merged**